third_party/nvidia/lib/Dialect/NVWS/Transforms/AssignStagePhase.cpp

```diff
@@ -218,24 +218,45 @@ template <class T> struct AssignStagePhase {
   StagePhase assignArefIndexInBlock(Block *block, StagePhase index) {
     for (auto &op : llvm::make_early_inc_range(*block)) {
       if (auto opT = isValidOp(&op)) {
-        ImplicitLocOpBuilder b(opT.getLoc(), opT);
-
-        auto nextStage = b.create<arith::AddIOp>(
-            index.stage, b.create<arith::ConstantIntOp>(1, 32));
+        ImplicitLocOpBuilder builder(opT.getLoc(), opT);
+        auto partitionIds = getPartitionIds(&op);
+        auto stageCluster = getStageCluster(&op);
+
+        auto createInto = [&](auto opTy, auto... args) {
+          using ty = decltype(opTy);
+          ty op = builder.create<ty>(std::forward<decltype(args)>(args)...);
+          if (partitionIds) {
+            // phase & stage computations are cloned to partition 0 as well
+            auto ids = *partitionIds;
+            ids.insert(0);
+            setPartition(op, ids);
+            if (stageCluster) {
+              op->setAttr(kLoopStageAttrName,
+                          builder.getI32IntegerAttr(stageCluster->first));
+              op->setAttr(kLoopClusterAttrName,
+                          builder.getI32IntegerAttr(stageCluster->second));
+            }
+          }
+          return op;
+        };
+
+        auto nextStage = createInto(arith::AddIOp{}, index.stage,
+                                    createInto(arith::ConstantIntOp{}, 1, 32));
         auto arefBuf = opT.getAref()
                            .template getDefiningOp<nvws::ArefCreateOp>()
                            .getOperand(0);
         auto depth = getArefDepth(cast<MemDescType>(arefBuf.getType()));
 
         auto cnd =
-            b.create<arith::CmpIOp>(arith::CmpIPredicate::eq, nextStage,
-                                    b.create<arith::ConstantIntOp>(depth, 32));
-        auto zero = b.create<arith::ConstantIntOp>(0, 32);
-        index.stage = b.create<arith::SelectOp>(cnd, zero, nextStage);
-
-        auto nextPhase = b.create<arith::XOrIOp>(
-            index.phase, b.create<arith::ConstantIntOp>(1, 32));
-        index.phase = b.create<arith::SelectOp>(cnd, nextPhase, index.phase);
+            createInto(arith::CmpIOp{}, arith::CmpIPredicate::eq, nextStage,
+                       createInto(arith::ConstantIntOp{}, depth, 32));
+        auto zero = createInto(arith::ConstantIntOp{}, 0, 32);
+        index.stage = createInto(arith::SelectOp{}, cnd, zero, nextStage);
+
+        auto nextPhase = createInto(arith::XOrIOp{}, index.phase,
+                                    createInto(arith::ConstantIntOp{}, 1, 32));
+        index.phase =
+            createInto(arith::SelectOp{}, cnd, nextPhase, index.phase);
```

Comment:
Don't we need to update the partition attribute of a parent op when we add a new partition to its body?

Reply:
> The partition for the if-op itself is then determined as the union of those output partitions

My point is that we should do that for each pass that adds a new op. We don't need a special-case convention for an if op (or for op), we can always say that all ops in a loop must be annotated with the correct partition.  

So why not do that as a post-process of assign-stage in this PR?